Crash ``'Call' object has no attribute 'name'`` when accessing a private class variable

### Bug description

Seems similar to #4638, but reproduces after its fix

When parsing the following file:

```python
class A:
    @classmethod
    def b(cls) -> None:
        cls.__a = ''

    def a(self):
        type(self).__a

```

pylint crashed with a ``AttributeError`` and with the following stacktrace:
```
Traceback (most recent call last):
  File "C:\Program Files\Python39\lib\site-packages\pylint\lint\pylinter.py", line 1034, in _check_files
    self._check_file(get_ast, check_astroid_module, file)
  File "C:\Program Files\Python39\lib\site-packages\pylint\lint\pylinter.py", line 1069, in _check_file
    check_astroid_module(ast_node)
  File "C:\Program Files\Python39\lib\site-packages\pylint\lint\pylinter.py", line 1203, in check_astroid_module
    retval = self._check_astroid_module(
  File "C:\Program Files\Python39\lib\site-packages\pylint\lint\pylinter.py", line 1250, in _check_astroid_module
    walker.walk(node)
  File "C:\Program Files\Python39\lib\site-packages\pylint\utils\ast_walker.py", line 75, in walk
    self.walk(child)
  File "C:\Program Files\Python39\lib\site-packages\pylint\utils\ast_walker.py", line 77, in walk
    callback(astroid)
  File "C:\Program Files\Python39\lib\site-packages\pylint\checkers\classes.py", line 901, in leave_classdef
    self._check_unused_private_attributes(node)
  File "C:\Program Files\Python39\lib\site-packages\pylint\checkers\classes.py", line 1008, in _check_unused_private_attributes
    and attribute.expr.name in {"cls", "self", node.name}
AttributeError: 'Call' object has no attribute 'name'
```

### Configuration

_No response_

### Command used

```shell
pylint bugale_test.py
```


### Pylint output

```shell
Exception on node <ClassDef.A l.1 at 0x19bf0e5c820> in file 'C:\Users\bugaleb\Documents\GitHub\winagent\SmartViz\AutomationTests\bugale_test.py'
Traceback (most recent call last):
  File "C:\Program Files\Python39\lib\site-packages\pylint\utils\ast_walker.py", line 77, in walk
    callback(astroid)
  File "C:\Program Files\Python39\lib\site-packages\pylint\checkers\classes.py", line 901, in leave_classdef
    self._check_unused_private_attributes(node)
  File "C:\Program Files\Python39\lib\site-packages\pylint\checkers\classes.py", line 1008, in _check_unused_private_attributes
    and attribute.expr.name in {"cls", "self", node.name}
AttributeError: 'Call' object has no attribute 'name'
bugale_test.py:1:0: F0001: Fatal error while checking 'bugale_test.py'. Please open an issue in our bug tracker so we address this. There is a pre-filled template that you can use in 'C:\Users\bugaleb\AppData\Local\pylint\pylint\Cache\pylint-crash-2021-12-21-00.txt'. (fatal)
```


### Expected behavior

Not to crash

### Pylint version

```shell
pylint 2.12.2
astroid 2.9.0
Python 3.9.5 (tags/v3.9.5:0a7dcbd, May  3 2021, 17:27:52) [MSC v.1928 64 bit (AMD64)]
```


### OS / Environment

Windows 11

### Additional dependencies

_No response_

--- a/ChangeLog
+++ b/ChangeLog
@@ -35,6 +35,11 @@ Release date: TBA
   Closes #4798
   Closes #5081
 
+* Fix a crash in ``unused-private-member`` checker when analyzing code using
+  ``type(self)`` in bound methods.
+
+  Closes #5569
+
 * Pyreverse - add output in mermaidjs format
 
 * ``used-before-assignment`` now considers that assignments in a try block

--- a/doc/whatsnew/2.13.rst
+++ b/doc/whatsnew/2.13.rst
@@ -74,6 +74,11 @@ Other Changes
 
   Closes #4716
 
+* Fix a crash in ``unused-private-member`` checker when analyzing code using
+  ``type(self)`` in bound methods.
+
+  Closes #5569
+
 * Fix crash in ``unnecessary-dict-index-lookup`` checker if the output of
   ``items()`` is assigned to a 1-tuple.
 

--- a/pylint/checkers/classes/class_checker.py
+++ b/pylint/checkers/classes/class_checker.py
@@ -970,6 +970,9 @@ a metaclass class method.",
                 if attribute.attrname != assign_attr.attrname:
                     continue
 
+                if self._is_type_self_call(attribute.expr):
+                    continue
+
                 if (
                     assign_attr.expr.name
                     in {
@@ -2034,11 +2037,19 @@ a metaclass class method.",
 
         Name is `self` for method, `cls` for classmethod and `mcs` for metaclass.
         """
-        return (
-            self._first_attrs
-            and isinstance(node, nodes.Name)
-            and node.name == self._first_attrs[-1]
-        )
+        if self._first_attrs:
+            first_attr = self._first_attrs[-1]
+        else:
+            # It's possible the function was already unregistered.
+            closest_func = utils.get_node_first_ancestor_of_type(
+                node, nodes.FunctionDef
+            )
+            if closest_func is None:
+                return False
+            if not closest_func.args.args:
+                return False
+            first_attr = closest_func.args.args[0].name
+        return isinstance(node, nodes.Name) and node.name == first_attr
 
 
 def _ancestors_to_call(klass_node, method="__init__"):

--- a/pylint/checkers/classes/class_checker.py
+++ b/pylint/checkers/classes/class_checker.py
@@ -1669,7 +1669,7 @@ a metaclass class method.",
             return False
         return frame_name and frame_name in PYMETHODS
 
-    def _is_type_self_call(self, expr):
+    def _is_type_self_call(self, expr: nodes.NodeNG) -> bool:
         return (
             isinstance(expr, nodes.Call)
             and isinstance(expr.func, nodes.Name)
@@ -2032,7 +2032,7 @@ a metaclass class method.",
         """
         return self._is_mandatory_method_param(node.expr)
 
-    def _is_mandatory_method_param(self, node):
+    def _is_mandatory_method_param(self, node: nodes.NodeNG) -> bool:
         """Check if nodes.Name corresponds to first attribute variable name
 
         Name is `self` for method, `cls` for classmethod and `mcs` for metaclass.

--- a/tests/functional/u/unused/unused_private_member.py
+++ b/tests/functional/u/unused/unused_private_member.py
@@ -309,3 +309,13 @@ class Foo:
 
     def method(self):
         print(self.__class__.__ham)
+
+
+class TypeSelfCallInMethod:
+    """Regression test for issue 5569"""
+    @classmethod
+    def b(cls) -> None:
+        cls.__a = ''  # [unused-private-member]
+
+    def a(self):
+        return type(self).__a

--- a/tests/functional/u/unused/unused_private_member.txt
+++ b/tests/functional/u/unused/unused_private_member.txt
@@ -17,3 +17,4 @@ unused-private-member:243:12:243:50:FalsePositive4681.__init__:Unused private me
 unused-private-member:274:4:275:26:FalsePositive4849.__unused_private_method:Unused private member `FalsePositive4849.__unused_private_method()`:UNDEFINED
 unused-private-member:291:4:294:44:Pony.__init_defaults:Unused private member `Pony.__init_defaults(self)`:UNDEFINED
 unused-private-member:296:4:298:20:Pony.__get_fur_color:Unused private member `Pony.__get_fur_color(self)`:UNDEFINED
+unused-private-member:318:8:318:15:TypeSelfCallInMethod.b:Unused private member `TypeSelfCallInMethod.__a`:UNDEFINED

